kotlin-mbedtls/src/main/kotlin/org/opencoap/ssl/transport/DtlsServer.kt

```diff
@@ -291,4 +311,14 @@ class DtlsServer(
             sessionContext = sessionContext.copy(authentication = authentication)
         }
     }
+
+    interface DtlsSessionLifecycleCallbacks {
+        enum class Reason {
+            SUCCEEDED, FAILED, CLOSED, EXPIRED
+        }
+        fun handshakeStarted(adr: InetSocketAddress, ctx: SslHandshakeContext) = Unit
+        fun handshakeFinished(adr: InetSocketAddress, ctx: SslHandshakeContext, reason: Reason, throwable: Throwable? = null) = Unit
+        fun sessionStarted(adr: InetSocketAddress, ctx: SslSession) = Unit
+        fun sessionFinished(adr: InetSocketAddress, ctx: SslSession, reason: Reason, throwable: Throwable? = null) = Unit
```

Comment:
Let's not expose `SslHandshakeContext` or `SslSession` those are classes that holds reference to native memory. (We should probably mark then as internal)

Reply:
Ok. Then we will need to pass some information to the callbacks directly.

But is it a problem if the memory references are internal and all the publicly available interfaces are read/only?

Reply:
Or can we expose `DtlsSession` and `DtlsHandshake` instead?